### PR TITLE
 Rewrite BUnionType's `getMemberTypes()` usages using SemTypes

### LINKED ISSUE #43344: [Improvement]: Refactor `BUnionType`'s  `getMemberTypes()` usages
### Description

$subject. We need to get rid of `getMemberTypes()` and `getOriginalMemberTypes()` usages in the current compiler, in order to do #43343.

This issue depends on #43126 which is in PR sent state now.

### LINKED ISSUE #43437: [Bug]: No error when `distinct` is used for an object and error type union
### Description

$subject.

```
type Y error|object {};

type X distinct Y; // This should give a compilation error.

public function main() {
    X x = error("hi");
    io:println(x);
}
```

> distinct-type-descriptor D, where D is distinct T
> The type T must be definite and must be a **subtype of object or a subtype of error**.

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject

Fixes #43344
Fixes #43437

## Approach
n/a

## Samples
n/a

## Remarks
n/a

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ServiceDesugar.java [hunk 1]
```diff
@@ -268,7 +268,7 @@ private void addMethodInvocation(Location pos, BLangSimpleVarRef varRef, BInvoka
         // call is generated in BIRGen. Casting to the first listener type should be fine as actual method invocation
         // is based on the value rather than the type.
         BType listenerType = getListenerType(varRef.getBType());
-        if (!types.isSameType(listenerType, varRef.getBType())) {
+        if (listenerType.tag != varRef.getBType().tag || !types.isSameType(listenerType, varRef.getBType())) {
```
Comment (heshanpadmasiri) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ServiceDesugar.java:271]: Can't we use `isSameTypeIncludingTags` here as well?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 2]
```diff
@@ -2271,25 +2272,7 @@ private void addErrorTypesToSet(BType returnType, LinkedHashSet<BType> errorType
...
-            for (BType memberType : ((BUnionType) returnType).getMemberTypes()) {
-                if (hasNonErrorType(memberType)) {
-                    return true;
-                }
-            }
-            return false;
-        }
-
-        return true;
+        return !Core.isEmpty(types.typeCtx(), Core.diff(returnType.semType(), PredefinedType.ERROR));
```
Comment (heshanpadmasiri) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:2275]: Isn't this same as `!Core.isSubtyp(returnType.semType(), PredefinedType.ERROR))`?  Also in the previous code we had handled the case where `returnType` could be null don't we need to handle it here a well?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 3]
```diff
@@ -3333,14 +3316,14 @@ public void visit(BLangCheckedExpr checkedExpr, AnalyzerData data) {
 
         BType exprType = Types.getImpliedType(enclInvokable.getReturnTypeNode().getBType());
         BType checkedExprType = checkedExpr.expr.getBType();
-        BType errorType = types.getErrorTypes(checkedExprType);
+        SemType errorType = types.getErrorIntersection(checkedExprType.semType());
 
-        if (errorType == symTable.semanticError) {
+        if (PredefinedType.NEVER.equals(errorType)) {
```
Comment (heshanpadmasiri) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:3321]: Shall we use `Core.isNever()` instead?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 4]
```diff
@@ -5720,21 +4017,9 @@ private boolean checkAttachMethod(BAttachedFunction func) {
...
-                return true;
-            }
-
-            if (type.tag != TypeTags.OBJECT) {
-                return false;
-            }
-
-            return Symbols.isService(type.tsymbol);
+            ObjectQualifiers quals = new ObjectQualifiers(false, false, ObjectQualifiers.NetworkQualifier.Service);
+            SemType serviceObjTy = new ObjectDefinition().define(typeEnv(), quals, new ArrayList<>(0));
```
Comment (heshanpadmasiri) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4021]: Same as isolatedObjTy as above

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java [hunk 5]
```diff
@@ -507,15 +511,17 @@ private void populateMemberSemTypes(BType memberType, LinkedHashSet<SemType> mem
...
     }
 
-    public SemType semTypeIgnoringTypeIds() {
-        populateMemberSemTypes(true);
-        return computeResultantUnion(memberSemTypes);
+    /**
+     * When the type is mutated we need to reset resolved semType.
+     */
+    public void resetSemType() {
+        this.semType = null;
```
Comment (heshanpadmasiri) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java:524]: Shouldn't we also clear the `memberSemTypes` as well?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java [hunk 6]
```diff
@@ -112,15 +115,15 @@ private boolean canOptimizeNilUnionCheck(BType sourceType, BType targetType) {
             return false;
         }
         boolean foundNil = false;
-        BType otherType = null;
-        for (BType bType : ((BUnionType) sourceType).getMemberTypes()) {
-            if (JvmCodeGenUtil.getImpliedType(bType).tag == TypeTags.NIL) {
+        SemType otherTy = null;
+        for (SemType s : ((BUnionType) sourceType).getMemberSemTypes()) {
+            if (PredefinedType.NIL.equals(s)) {
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java]: Fixed in [2666d57](https://github.com/ballerina-platform/ballerina-lang/pull/43341/commits/2666d570f083357ec419ae22ced34ac01ca9376f)

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java [hunk 7]
```diff
@@ -137,16 +140,17 @@ private boolean canOptimizeErrorCheck(BType sourceType, BType targetType) {
         if (targetType.tag != TypeTags.ERROR || sourceType.tag != TypeTags.UNION) {
             return false;
         }
-        BType errorType = null;
+        SemType errorTy = null;
         int foundError = 0;
-        for (BType bType : ((BUnionType) sourceType).getMemberTypes()) {
-            if (bType.tag == TypeTags.ERROR) {
+        for (SemType s : ((BUnionType) sourceType).getMemberSemTypes()) {
```
[the remark this answers is not in the document]
Comment (lochana-chathura) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTypeTestGen.java]: Thanks. Fixed in [99473f9](https://github.com/ballerina-platform/ballerina-lang/pull/43341/commits/99473f91fb13020c846406afe568b5f14d79970b)

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 2 above)
Comment (lochana-chathura) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Fixed the former. I think the null check we had previously is redundant/obsolete. No test for that. Hence, removed.
 

Screenshot 2024-11-11 at 18 15 07

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java (hunk 5 above)
Comment (lochana-chathura) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java:518]: We repopulate the memberSemTypes here, whenever the semtype is `null`.

https://github.com/ballerina-platform/ballerina-lang/blob/efc16e5a61cbfa9ce2402527d1a34bba934ef2d6/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java#L485-L492

**Pull request merged**